### PR TITLE
Add annotations for ref fields to public API surface

### PR DESCRIPTION
**Pull request by AaronRobinsonMSFT**
This adds the necessary annotations to facilitate updating to C# 11. They will all be removed and converted to use the `scoped` keyword once we update to the latest Roslyn with `ref` field support.

The desire is to also backport this to Preview 6.

/cc @stephentoub @jaredpar @cston

### CODE REVIEW

**Review by cston — APPROVED** [review]

File: src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/Unsafe.cs [hunk 1]
```diff
@@ -671,7 +671,7 @@ public static void Write<T>(void* destination, T value)
         // Mono:AsRef
         [NonVersionable]
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public static ref T AsRef<T>(in T source)
+        public static ref T AsRef<T>([LifetimeAnnotation(true, false)] in T source)
```
Comment (jaredpar) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/Unsafe.cs:674]: FYI: this is effectively encoding that `AsRef` is the single method for stripping all safety from a `ref` value. It now removes `in` (changes mutability) and `scoped` (changes lifetime). I'm 100% okay with this, it's the design I prefer, however IIRC there was some debate on whether we should do this or have a separate method for each type of safety we strip. Wanted to double check that this debate was had and we're not accidentally just making that decision here.

**Review by jaredpar — APPROVED** [review]

File: src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/Unsafe.cs (hunk 1 above)
Comment (AaronRobinsonMSFT) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/Unsafe.cs:674]: @jkotas Do you have any thoughts on @jaredpar's recollection?

File: src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/MemoryMarshal.cs [hunk 2]
```diff
@@ -77,13 +77,13 @@ public static ReadOnlySpan<byte> AsBytes<T>(ReadOnlySpan<T> span)
...
         /// </summary>
-        public static ref T GetReference<T>(Span<T> span) => ref span._reference.Value;
+        public static ref T GetReference<T>([LifetimeAnnotation(false, true)] Span<T> span) => ref span._reference.Value;
 
         /// <summary>
         /// Returns a reference to the 0th element of the ReadOnlySpan. If the ReadOnlySpan is empty, returns a reference to the location where the 0th element
         /// would have been stored. Such a reference may or may not be null. It can be used for pinning but must never be dereferenced.
         /// </summary>
-        public static ref T GetReference<T>(ReadOnlySpan<T> span) => ref span._reference.Value;
+        public static ref T GetReference<T>([LifetimeAnnotation(false, true)] ReadOnlySpan<T> span) => ref span._reference.Value;
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/MemoryMarshal.cs:86]: Are the above two annotations needed?  Or could any callers that need the safety stripped away use Unsafe.AsRef as the sole escape hatch?

Comment (AaronRobinsonMSFT) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/MemoryMarshal.cs:86]: Let me play with this a bit.

Comment (AaronRobinsonMSFT) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/MemoryMarshal.cs:86]: This isn't needed. Should I remove here and in the Preview 6 PR or just leave this as is for now? We won't need this in Preview 7.

Comment (AaronRobinsonMSFT) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/MemoryMarshal.cs:86]: Will leave here. Fixed in https://github.com/dotnet/runtime/tree/feature/csharp_ref_fields.

**Comment by AaronRobinsonMSFT** [comment]
Wasm failure (1) is acknowledged at https://github.com/dotnet/runtime/pull/70991#issuecomment-1160878494. Failure (2) is https://github.com/dotnet/runtime/issues/71096.

**Review by stephentoub — APPROVED** [review]
Assuming this matches what was merged for P6, LGTM

File: src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/Unsafe.cs (hunk 1 above)
Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/Unsafe.cs:674]: The main concern on my end is making unsafe code harder to reason about. There are both pros and cons to having a single method here and I think its overall fine for this scenario.

I think the bigger annoyance on my end is actually the IL bloat and readability issues around having to strip `in` when doing more complex `Unsafe` logic, even where the underlying value isn't mutated, but that's a separate issue and one that is ideally handled by the planned language to allow migration from `ref` to `in` and `ref readonly` without being a binary or source breaking change.

**Pull request merged**